### PR TITLE
AWS - Add set-output-lifecycle action for bedrock-evaluation-job

### LINKED ISSUE #11072: AWS - Add a remediation action for Bedrock evaluation-job output lifecycle retention
### Describe the feature

`aws.bedrock-evaluation-job` has an `output-retention` filter that resolves a job's S3 output location and reports whether an enabled lifecycle expiration rule covers it, but no action pairs with it. The fix is a lifecycle expiration rule on the job's output bucket, scoped to its output prefix, and nothing adds one automatically.

Add an action, on `aws.bedrock-evaluation-job` or as a documented related-resource pattern against `aws.s3`, that gives a matched job's output bucket an enabled lifecycle expiration rule covering the job's output prefix within a configured number of days. The action should add a rule only when the prefix isn't already covered, and must not remove or alter unrelated existing rules on the same bucket.

This gap blocks enforcement only; `inform` already works through the existing `output-retention` filter.

### Extra information or context

- Resource: `aws.bedrock-evaluation-job` (`c7n/resources/bedrock.py`)
- Existing filter: `output-retention` (`BedrockEvaluationOutputRetention`), which already resolves the job's bucket, prefix, and matched lifecycle rules via `get_bedrock_output_lifecycle`.
- Target API: `PutBucketLifecycleConfiguration` (https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketLifecycleConfiguration.html)
- Correctness requirement: the added or updated rule must scope to the same prefix the `output-retention` filter already computes (`c7n:BedrockEvaluationOutput.ArtifactPrefix`), not the whole bucket.

### PR DESCRIPTION
**Pull request by mrinalpravi**
### Changes

* Added `get_bedrock_job_id()` to extract job IDs from Bedrock job ARNs.
* Added `set-output-lifecycle` action to configure per-job S3 output retention.

  * Reuses existing output URI/prefix resolution logic.
  * Preserves unrelated bucket lifecycle rules.
  * Idempotently skips already-covered prefixes and safely handles inaccessible buckets.
  * Declares required S3 lifecycle/versioning permissions.
* Added 8 unit tests covering permissions, invalid/missing buckets, existing rules, updates, idempotency, and shared buckets.

**Tests Verified:**
Screenshot 2026-09-08 at 12 08 49 PM

**Example:**

```yaml
policies:
  - name: bedrock-evaluation-output-retention-remediate
    resource: aws.bedrock-evaluation-job
    filters:
      - type: output-retention
        op: gt
        value: 30
    actions:
      - type: set-output-lifecycle
        days: 30
```

Fixes #11072

### CODE REVIEW

File: tests/test_bedrock.py [hunk 1]
```diff
@@ -1316,6 +1316,203 @@ def test_permissions_and_validation(self):
...
+                'c7n:DeniedMethods': ['get_bucket_lifecycle_configuration']},
+        }
+        action._get_buckets = mock.Mock(return_value=buckets)
+        jobs = [
+            self.job('a', 'id-a', 'missing', 'evaluations'),
+            self.job('b', 'id-b', 'denied', 'evaluations'),
+        ]
+        action.process(jobs)  # should not raise, nothing to put
+
+    def test_already_covered_skips_put(self):
```
Comment (KodyMullins) [inline tests/test_bedrock.py:1368]: Mock test go against the project testing [conventions](https://github.com/cloud-custodian/cloud-custodian/blob/7ec402714821b35ce223fb3f58d7f1f545119920/docs/source/developer/tests.rst#L4)

Please update all relevant test to follow the record/replay test pattern to make live api calls to a aws instance

File: c7n/resources/bedrock.py [hunk 2]
```diff
@@ -1245,6 +1245,132 @@ def process(self, resources, event=None):
...
+            if effective is not None and effective >= days:
+                continue
+
+            rule_id = 'c7n-bedrock-output-retention-%s' % get_bedrock_job_id(resource)
+            new_rule = {
+                'ID': rule_id,
+                'Status': 'Enabled',
+                'Filter': {'Prefix': artifact_prefix},
+                'Expiration': {'Days': days},
+            }
```
Comment (KodyMullins) [inline c7n/resources/bedrock.py:1358]: This code block does not account for buckets that have versioning enabled. 
Any old file versions would there for never be cleaned up only the current version. 

``` suggestion
new_rule = {
                'ID': rule_id,
                'Status': 'Enabled',
                'Filter': {'Prefix': artifact_prefix},
                'Expiration': {'Days': days},
            }
            if _bucket_is_versioned(versioning):
                new_rule['NoncurrentVersionExpiration'] = {'NoncurrentDays': days}
```

File: tests/test_bedrock.py [hunk 3]
```diff
@@ -1316,6 +1316,203 @@ def test_permissions_and_validation(self):
             self.get_filter({'key': 'Versioning.Status', 'value': 'Enabled'})
 
 
+class TestSetBedrockEvaluationOutputLifecycle(BaseTest):
```
Comment (KodyMullins) [inline tests/test_bedrock.py:1319]: The test class does not currently produce enough code coverage for ci to pass.

Image

File: c7n/resources/bedrock.py [hunk 4]
```diff
@@ -1245,6 +1245,132 @@ def process(self, resources, event=None):
...
+        assembler = BedrockOutputBucketAssembly(self.manager)
+        assembler.initialize()
+        assembler.augment_fields = {'Location', 'Lifecycle', 'Versioning'}
+        buckets = {}
+        for name in bucket_names:
+            bucket = assembler.assemble({'Name': name})
+            if name in assembler.not_found_buckets:
+                bucket['c7n:BedrockOutputBucketError'] = 'bucket-not-found'
+            buckets[name] = bucket
+        return buckets
```
Comment (KodyMullins) [inline c7n/resources/bedrock.py]: This code block is almost an exact match for lines 1176-1189.

It would probably benefit both to have a shared helper to prevent drift between the two actions

```
def assemble_bedrock_output_buckets(manager, bucket_names, augment_fields):
    if not bucket_names:
        return {}
    assembler = BedrockOutputBucketAssembly(manager)
    assembler.initialize()
    assembler.augment_fields = augment_fields
    buckets = {}
    for name in bucket_names:
        bucket = assembler.assemble({'Name': name})
        if name in assembler.not_found_buckets:
            bucket['c7n:BedrockOutputBucketError'] = 'bucket-not-found'
        buckets[name] = bucket
    return buckets

```

File: c7n/resources/bedrock.py [hunk 5]
```diff
@@ -1245,6 +1245,132 @@ def process(self, resources, event=None):
         return results
 
 
+def get_bedrock_job_id(resource):
```
Comment (KodyMullins) [inline c7n/resources/bedrock.py]: This should be abstracted  to a shared helper in `class BedrockOutputBucketAssembly(BucketAssembly):`

```
def _bedrock_job_id_from_arn(job_arn):
    return job_arn.rsplit('/', 1)[-1] if '/' in job_arn else None


def get_bedrock_output_artifact_prefix(output_prefix, resource):
    """Return the common prefix under which Bedrock writes a job's artifacts."""
    job_name = resource.get('jobName')
    job_id = _bedrock_job_id_from_arn(resource.get('jobArn', ''))
```

Then 
```
rule_id = 'c7n-bedrock-output-retention-%s' % get_bedrock_job_id(resource)
```

Can be replaced with 

```
            job_id = _bedrock_job_id_from_arn(resource.get('jobArn', '')) or resource.get(
                'jobName', '')
            rule_id = 'c7n-bedrock-output-retention-%s' % job_id

```
To prevent duplication of id extraction logic.

**Review by KodyMullins — CHANGES_REQUESTED** [review]

File: tests/test_bedrock.py (hunk 1 above)
Comment (mrinalpravi) [reply tests/test_bedrock.py:1368]: Addressed

File: c7n/resources/bedrock.py (hunk 2 above)
Comment (mrinalpravi) [reply c7n/resources/bedrock.py:1358]: Addressed

File: c7n/resources/bedrock.py (hunk 4 above)
Comment (mrinalpravi) [reply c7n/resources/bedrock.py]: addressed

Comment (mrinalpravi) [reply c7n/resources/bedrock.py]: addressed

File: c7n/resources/bedrock.py (hunk 5 above)
Comment (mrinalpravi) [reply c7n/resources/bedrock.py]: addressed

File: tests/test_bedrock.py (hunk 3 above)
Comment (mrinalpravi) [reply tests/test_bedrock.py:1319]: addressed

File: c7n/resources/bedrock.py (hunk 5 above)
Comment (mrinalpravi) [reply c7n/resources/bedrock.py]: addresses

File: c7n/resources/bedrock.py (hunk 4 above)
Comment (mrinalpravi) [reply c7n/resources/bedrock.py]: addressed

File: c7n/resources/bedrock.py
Comment (KodyMullins) [inline c7n/resources/bedrock.py:1330]: missing a warning for the versioning that was added 
``` suggestion
 if 'get_bucket_lifecycle_configuration' in bucket.get('c7n:DeniedMethods', ()):
            denied = bucket.get('c7n:DeniedMethods', ())
            if 'get_bucket_lifecycle_configuration' in denied:
                self.log.warning(
                    "access denied reading lifecycle for bucket %s, skipping", bucket_name)
                continue
            if 'get_bucket_versioning' in denied:
                self.log.warning(
                    "access denied reading versioning for bucket %s, skipping", bucket_name)
                continue

```

File: c7n/resources/bedrock.py [hunk 6]
```diff
@@ -1346,13 +1337,17 @@ def _process_bucket(self, session, bucket, jobs):
...
             new_rule = {
                 'ID': rule_id,
                 'Status': 'Enabled',
                 'Filter': {'Prefix': artifact_prefix},
                 'Expiration': {'Days': days},
             }
+            if _bucket_is_versioned(versioning):
+                new_rule['NoncurrentVersionExpiration'] = {'NoncurrentDays': days}
             for index, existing in enumerate(rules):
                 if existing.get('ID') == rule_id:
```
Comment (KodyMullins) [inline c7n/resources/bedrock.py:1362]: Ai pointed this out to me.

Possible drift risk in the existing-rule comparison 

This does a full dict equality check between the AWS-returned existing rule and our locally-built new_rule. Per the S3 PutBucketLifecycleConfiguration/LifecycleRule API docs, a rule can carry additional fields we never set — e.g. AbortIncompleteMultipartUpload, or Expiration.ExpiredObjectDeleteMarker — and AWS doesn't merge/normalize rules across a PUT, so whatever's stored is returned verbatim on the next GET. If a rule under our c7n-bedrock-output-retention-<job_id> ID ever ends up with such an extra field (e.g. someone hand-edits it, or a future version of this action adds a field), this equality check will never match and we'll call put_bucket_lifecycle_configuration on every single run even though the retention guarantee is already satisfied.

In practice the blast radius is small — since the ID is per-job and effectively owned exclusively by this action, nothing else should be writing under it, and the one real recorded API response we have for this rule shape shows no injected extras. So this isn't causing a bug today. But it's a bit fragile as written; worth considering a narrower comparison of just the fields we actually manage (Status, Filter, Expiration.Days, NoncurrentVersionExpiration.NoncurrentDays) instead of whole-dict equality, so the action stays idempotent even if the stored rule ever picks up unrelated fields. Not blocking — flagging for you to decide if it's worth hardening now or leaving as a known edge case.

File: tests/test_bedrock.py [hunk 7]
```diff
@@ -1347,75 +1347,43 @@ def job(self, name, job_id, bucket, prefix):
...
 
     def test_skips_job_with_invalid_uri(self):
-        action = self.get_action()
-        action._get_buckets = mock.Mock(return_value={})
+        session_factory = self.replay_flight_data(
+            'bedrock_set_output_lifecycle_invalid_uri')
+        action = self.get_action(session_factory=session_factory)
+        # no bucket can be resolved from an invalid uri, so no s3 calls are
+        # made; a missing flight data fixture would fail this test.
         action.process([{'jobArn': 'arn:aws:bedrock:r:a:evaluation-job/id'}])
```
Comment (KodyMullins) [inline tests/test_bedrock.py:1355]: we could add assertions based on the log outputs since warnings are thrown for this test and the next one. 

```
with self.assertLogs('custodian.actions', level='WARNING') as cm:
    action.process([{'jobArn': 'arn:aws:bedrock:r:a:evaluation-job/id'}])
assert len(cm.output) == 1
assert 'missing-uri' in cm.output[0]
assert 'arn:aws:bedrock:r:a:evaluation-job/id' in cm.output[0]

```

**Review by KodyMullins — CHANGES_REQUESTED** [review]

**Comment by KodyMullins** [comment]
Thanks for addressing my concerns so quickly. I looks much better now.

**Comment by mrinalpravi** [comment]
Addressed review comments @KodyMullins

**Review by KodyMullins — APPROVED** [review]
